File: common/src/main/java/juuxel/adorn/networking/SetTradeStackC2SMessage.java
package juuxel.adorn.networking;

import juuxel.adorn.AdornCommon;
import net.minecraft.network.RegistryFriendlyByteBuf;
import net.minecraft.network.codec.ByteBufCodecs;
import net.minecraft.network.codec.StreamCodec;
import net.minecraft.network.protocol.common.custom.CustomPacketPayload;
import net.minecraft.world.item.ItemStack;

public record SetTradeStackC2SMessage(int syncId, int slotId, ItemStack stack) implements CustomPacketPayload {
    public static final Type<SetTradeStackC2SMessage> ID = new Type<>(AdornCommon.id("set_trade_stack"));
    public static final StreamCodec<RegistryFriendlyByteBuf, SetTradeStackC2SMessage> PACKET_CODEC = StreamCodec.composite(
        ByteBufCodecs.VAR_INT,
        SetTradeStackC2SMessage::syncId,
        ByteBufCodecs.VAR_INT,
        SetTradeStackC2SMessage::slotId,
        ItemStack.STREAM_CODEC,
        SetTradeStackC2SMessage::stack,
        SetTradeStackC2SMessage::new
    );

    @Override
    public Type<? extends CustomPacketPayload> type() {
        return ID;
    }
}

File: forge/src/main/kotlin/juuxel/adorn/platform/forge/networking/SetTradeStackC2SMessage.kt
package juuxel.adorn.platform.forge.networking

import net.minecraft.item.ItemStack
import net.minecraft.network.PacketByteBuf
import net.minecraft.network.packet.CustomPayload
import net.minecraft.util.Identifier

data class SetTradeStackC2SMessage(val syncId: Int, val slotId: Int, val stack: ItemStack) : CustomPayload {
    override fun id(): Identifier = AdornNetworking.SET_TRADE_STACK

    override fun write(buf: PacketByteBuf) {
        buf.writeVarInt(syncId)
        buf.writeVarInt(slotId)
        buf.writeItemStack(stack)
    }

    companion object {
        fun fromPacket(buf: PacketByteBuf): SetTradeStackC2SMessage =
            SetTradeStackC2SMessage(
                buf.readVarInt(),
                buf.readVarInt(),
                buf.readItemStack()
            )
    }
}
